File: submit-api/src/submit_api/models/project.py
"""Account Project model class.

Manages the account project
"""
from __future__ import annotations

from sqlalchemy import Column, event
from sqlalchemy.exc import ArgumentError, IntegrityError

from submit_api.exceptions import PermissionDeniedError

from .db import db


class Project(db.Model):
    """Definition of the Project entity."""

    __tablename__ = 'projects'

    id = Column(db.Integer, primary_key=True, autoincrement=True)
    name = Column(db.String(), nullable=False)
    proponent_id = Column(db.Integer(), nullable=False, unique=True)
    proponent_name = Column(db.String(), nullable=False)
    ea_certificate = Column(db.String(255), nullable=True, default=None)

    def __init__(self, **kwargs):
        """Initialize the Project entity."""
        raise ArgumentError("Project is read-only, cannot create new instances.")

    @classmethod
    def get_all_projects_in_ids(cls, project_ids):
        """Get all projects in the given project ids."""
        return cls.query.filter(cls.id.in_(project_ids)).all()


@event.listens_for(Project, 'before_insert')
def before_insert():
    """Raise an error when trying to insert into the Project table."""
    raise IntegrityError(
        "Insertions are not allowed on this table",
        params=None,
        orig=PermissionDeniedError('Insertions are not allowed on this table')
    )


@event.listens_for(Project, 'before_update')
def before_update():
    """Raise an error when trying to update the Project table."""
    raise IntegrityError(
        "Updates are not allowed on this table",
        params=None,
        orig=PermissionDeniedError('Updates are not allowed on this table')
    )


@event.listens_for(Project, 'before_delete')
def before_delete():
    """Raise an error when trying to delete from the Project table."""
    raise IntegrityError(
        "Deletions are not allowed on this table",
        params=None,
        orig=PermissionDeniedError('Deletions are not allowed on this table')
    )

File: submit-web/src/models/Project.ts
import { SubmissionPackage } from "./Package";

export type Project = {
  id: number;
  name: string;
  proponent_id: number;
  proponent_name: string;
  ea_certificate?: string;
};

export type AccountProject = {
  id: number;
  project_id: number;
  account_id: number;
  project: Project;
  packages: SubmissionPackage[];
};

export const createDefaultAccountProject = (): AccountProject => ({
  id: 0,
  project_id: 0,
  account_id: 0,
  project: {
    id: 0,
    name: "",
    proponent_id: 0,
    proponent_name: "",
  },
  packages: [],
});
